### PR TITLE
DRILL-8235: Add Storage Plugin for Google Sheets

### PR DESCRIPTION
**Pull request by cgivre**
# [DRILL-8235](https://issues.apache.org/jira/browse/DRILL-8235): Add Storage Plugin for Google Sheets

## Description
Google Sheets is a very commonly used data source among business users.  Presto and other query engines do include integrations with Google Sheets and so it would be useful for Drill to add this functionality. 

The proposed plugin supports both reading and writing to Google Sheets. 

Here's a quick demo: 
https://www.youtube.com/watch?v=pS2FKxrluBA

This PR makes a few addition minor modifications that are not directly related to Google Sheets.

* Modification to the OAuth token storage to allow plugins besides HTTP to use Drill's token storage mechanism
* Modification to the OAuth token storage to store the token expiration time.
* Reapplication of `.gitignore` and removing some files which should have been ignored.  Note this caused a unit test to break, so I had to rename a file to fix this.
* This PR also integrates [DRILL-8270](https://issues.apache.org/jira/browse/DRILL-8270)

## Documentation
Docs in README.

- [x] Finalize Unit Tests
- [x] Fix writer
- [x] Remove logback file.
- [x] Code cleanup
- [x] Fix LGTM bugs

## Testing
Manually tested and added unit tests.

### CODE REVIEW

**Comment by jnturton** [comment]
Nice demo vid! Before I dive in, I see that a review has been requested but the PR is also marked draft. Can I switch it over to reviewable? Btw, I like the sheet specification syntax here and it's something I wondered about when doing a bug fix on the Excel format plugin recently. But I guess Excel being a format plugin, this syntax is not available and we have to stick with the format config opt that we have?

**Comment by cgivre** [comment]
> 

Thanks @jnturton !!  You certainly may start reviewing if you like.  I left it as draft for the moment as there is still some minor cleanup and debugging that I have to do.  I don't anticipate any major changes, but I'm adding more unit tests and fixing all the LGTM / CodeQL issues and stuff like that.  If you want to start... please feel free.

**Comment by cgivre** [comment]
@jnturton This is now ready for review.

File: contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsQueries.java
Comment (jnturton) [inline contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsQueries.java:136]: Isn't it necessary to look in the plan string for the projected columns to test the pushdown part? I would expect this test to pass even if there was no projection pushdown...

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsBatchReader.java
Comment (jnturton) [inline contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsBatchReader.java:27]:
```suggestion
import org.apache.drill.exec.physical.impl.scan.v3.ManagedReader;
```
Can we use the EVF2 API here?

**Review by jnturton — CHANGES_REQUESTED** [review]

File: contrib/native/client/patches/zookeeper-3.4.6-x64.patch
Comment (cgivre) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: I don't know.  Patch files are supposed to be excluded by the .gitignore.  What happened was I reapplied Drill's `.gitignore` and it deleted that patch file.   I would assume this file should not be included.

Comment (jnturton) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: It's very old:
```
commit d944918ff8a1f76d9b4c400512a130bff7fb2c5f
Author: Parth Chandra <pchandra@maprtech.com>
Date:   Wed Jun 11 18:10:34 2014 -0700

    DRILL-1021: Windows build
```
@vdiravka any thoughts about this patch file?

File: .gitignore
Comment (cgivre) [reply .gitignore:32]: Fixed.

File: contrib/storage-googlesheets/README.md
Comment (cgivre) [reply contrib/storage-googlesheets/README.md:55]: Fixed

File: contrib/storage-googlesheets/README.md
Comment (cgivre) [reply contrib/storage-googlesheets/README.md:65]: Fixed here and elsewhere.

File: contrib/storage-googlesheets/README.md
Comment (cgivre) [reply contrib/storage-googlesheets/README.md:95]: Thx!!

File: contrib/storage-googlesheets/README.md
Comment (cgivre) [reply contrib/storage-googlesheets/README.md:103]: Removed... I was going to do this, but I think `allTextMode` will suffice.

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsPushDownListener.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsPushDownListener.java:98]: For now yes.  If we can figure out how to use the Google SDK for filter pushdowns, we should uncomment out that line.  I'll add a comment there, explaining.  I'll do my best to keep the snark factor low, but no promises.

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsStoragePlugin.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsStoragePlugin.java:165]: Fixed.

File: contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsLimitPushdown.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsLimitPushdown.java:42]: Fixed, here and elsewhere.

File: contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsQueries.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsQueries.java:136]: Fixed.

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsGroupScan.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsGroupScan.java:301]: I'll add a TODO.  I think you are correct that we can get the max row count, but I'll save that for future work.

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/utils/GoogleSheetsTypifier.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/utils/GoogleSheetsTypifier.java]: Moved to `common`.

File: contrib/storage-googlesheets/README.md
Comment (cgivre) [reply contrib/storage-googlesheets/README.md]: Fixed

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsGroupScan.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsGroupScan.java:301]: Added TODO.

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsBatchReader.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsBatchReader.java:27]: @jnturton It turns out that EVF2 is missing a few bits to fully convert storage plugins.  There isn't a ManagedReader class for non-file systems.  I'd be happy to do that, but if it's ok with you, could we make that a separate PR?

Comment (jnturton) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsBatchReader.java:27]: > There isn't a ManagedReader class for non-file systems. I'd be happy to do that, but if it's ok with you, could we make that a separate PR?

Sounds good, thanks.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/security/OAuthTokenCredentials.java
Comment (cgivre) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/security/OAuthTokenCredentials.java]: Good catch!  I had been working on this for some time, and I think I created that before we merged the final version of OAuth support.  In any event, the duplicate class has been removed.

File: contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/columns/GoogleSheetsDateColumnWriter.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/columns/GoogleSheetsDateColumnWriter.java]: Done!

File: contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsWriter.java
Comment (cgivre) [reply contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsWriter.java]: I removed this and added a comment about it.  With the current implementation, when you execute a CTAS query, you do so with the desired file name. IE:

```sql
CREATE TABLE googlesheets.myfile AS
SELECT <stuff>
FROM <data>
```

The problem occurs when you go to test that and you need the file ID.  Google's SDK doesn't give you a way that I found to go from file name to file ID.

File: logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java
Comment (cgivre) [reply logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java:37]: I can add this to the generic FormatPlugin super class.  Should that be a separate PR?

**Comment by cgivre** [comment]
@jnturton Thank you very much for your review of this beast.  I believe I have addressed all your comments, as well as fixed the final bugs with the writer.

File: logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java
Comment (jnturton) [reply logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java:37]: Yes, thank you, let's make case insensitive properties its own PR. Please revert it here.

File: contrib/native/client/patches/zookeeper-3.4.6-x64.patch
Comment (jnturton) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: If this PR really wants to be the one to delete this file from version control then I think it should include some comment  in the description like "this patch was added for X and is no longer relevant". I can see that it's a C code patch to ZooKeeper with a title of `[PATCH] Allow zookeeper to build in x64` which makes me 99% sure it's completely obsolete. @vdiravka, @vvysotskyi any advice for us here?

**Review by jnturton — CHANGES_REQUESTED** [review]
Thanks for all the fixes. Just two minor items remaining from me.

File: logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java
Comment (cgivre) [reply logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java]: Done!

File: contrib/native/client/patches/zookeeper-3.4.6-x64.patch
Comment (vdiravka) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: There is [readme.win.txt](https://github.com/apache/drill/blob/master/contrib/native/client/readme.win.txt#L135) about it:
> The 3.4.6 release of Zookeeper does not build correctly on 64 bit windows. To fix that for the 64 bit build, apply patch zookeeper-3.4.6-x64.patch

Comment (cgivre) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: @vdiravka Forgive my lack of knowledge here, but the current `.gitignore` excludes patch files.  Is there a way to ignore all patch files, but leave that one?  How would you suggest we fix this?

@jnturton Aside from this patch file, are we good to go, pending CI?

Comment (jnturton) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: I think the .gitignore is working properly already because all of our previous commits have left zookeeper-3.4.6-x64.patch in peace. How about we do it this way? If this PR wants to delete the patch then it should simultaneously update readme.win.txt loosely saying "ZooKeeper 3.4.6 is ancient, ZooKeeper 3.5.7 is current and does not need any patch to build on Windows so the patch remains in the Git history only". Alternatively, we revert the deletion here and carry the cruft a bit longer.

+1 one from me otherwise.

Comment (cgivre) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: Sounds good.  Updating README.

**Review by jnturton — APPROVED** [review]

File: contrib/native/client/patches/zookeeper-3.4.6-x64.patch
Comment (cgivre) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: Done!

Comment (vdiravka) [reply contrib/native/client/patches/zookeeper-3.4.6-x64.patch:1]: @cgivre In any case, if the issue persists, there is a [way](https://stackoverflow.com/questions/987142/make-gitignore-ignore-everything-except-a-few-files) how to exclude only one file from the pattern:
`!contrib/native/client/patches/zookeeper-3.4.6-x64.patch`

**Pull request merged**
